Add Iterators.first and Iterators.last which don't throw for empty collections

non-Jedi opened the pull request:
EDIT: This PR now implements option 2 discussed at the bottom of this post rather than option 1 which it originally implemented.

On zulip, there was a thread about the fact that there's not a good
way of finding the first item matching a predicate in a non-indexable iterable without
manually writing out the `for`-loop
(<https://julialang.zulipchat.com/#narrow/stream/225540-gripes/topic/.60find.60.20in.20iterables>). One
possibility is using the existing `first` function composed with
`Iterators.filter`:

    first(Iterators.filter(>(5), 1:10))

The problem with this approach is that it requires wrapping in a `try`
block if the predicate not being satisfied is a
possibility (`Iterators.filter` returns an empty collection). There
should be a way of expressing this without using exception-handling as
control flow or writing the `for`-loop manually.

Ideally, I think first and last should just return
`Union{Some{T},Nothing}` always, but that would be a breaking
change. So I see two obvious alternatives:

1. This commit adds an optional predicate argument to `first` and `last`
   which is mapped over the collection and the first/last value
   satisfying predicate is returned wrapped in `Some`.
2. Add a `Iterators.first`/`Iterators.last` which are identical to the
   versions in Base except that when successful they return `Some{T}`
   and when the collection is empty return `Nothing`

non-Jedi committed:
Add methods of first and last accepting a predicate

On zulip, there was a thread about the fact that there's not a good
way of finding the first item matching a predicate in a non-indexable
iterable without manually writing out the for-loop. One possibility is
using the existing first function composed with Iterators.filter:

    first(Iterators.filter(>(5), 1:10))

The problem with this approach is that it requires wrapping in a try
block if the predicate not being satisfied is a
possibility (Iterators.filter returns an empty collection). There
should be a way of expressing this without using exception-handling as
control flow or writing the for-loop manually.

Ideally, I think first and last should just return
Union{Some{T},Nothing} always, but that would be a breaking
change. So I see two obvious alternatives:

1. This commit adds an optional predicate argument to first and last
   which is mapped over the collection and the first/last value
   satisfying predicate is returned.
2. Add a Iterators.first/Iterators.last which are identical to the
   versions in Base except that when successful they return Some{T}
   and when unsuccesful return Nothing

KristofferC commented on base/abstractarray.jl:
```diff
@@ -343,6 +343,34 @@ function first(itr)
+function first(predicate, itr)
+    for x in itr
+        predicate(x) && return Some(x)
+    end
```
```suggestion
    end
    return nothing
```

non-Jedi replied:
Is there a style guide for Julia base somewhere? I didn't know it was preferred to always use `return`, and I don't see it listed in https://docs.julialang.org/en/v1/manual/style-guide/

ararslan replied:
It's just a bit iffy to rely on the fact that `for` loops return `nothing`, so it's better to be explicit in the case that the loop completes.

KristofferC commented on base/abstractarray.jl:
```diff
@@ -388,6 +416,43 @@ julia> last([1; 2; 3; 4])
+    for x in itr
+        out = ifelse(predicate(x), Some(x), out)
+    end
+    out
```
```suggestion
    return out
```

KristofferC commented on base/abstractarray.jl:
```diff
@@ -388,6 +416,43 @@ julia> last([1; 2; 3; 4])
+function last(predicate, a::AbstractArray)
+    @inbounds for i in reverse(eachindex(a))
+        predicate(a[i]) && return Some(a[i])
+    end
```
```suggestion
    end
    return nothing
```

non-Jedi committed:
Use return keyword instead of implicit return

Co-authored-by: Kristoffer Carlsson <kristoffer.carlsson@chalmers.se>

ararslan commented on base/abstractarray.jl:
```diff
@@ -388,6 +417,44 @@ julia> last([1; 2; 3; 4])
+"""
+    last(predicate, coll)
+
+Get the last element of `coll` satisfying `predicate` wrapped in `Some`.
```
```suggestion
Get the last element of `coll` satisfying `predicate` wrapped in [`Some`](@ref).
```

ararslan commented on base/abstractarray.jl:
```diff
@@ -388,6 +417,44 @@ julia> last([1; 2; 3; 4])
+julia> last(<(5), 1:10)
+Some(4)
+
+julia> last(isodd, 2:2:10) |> isnothing
```
```suggestion
julia> isnothing(last(isodd, 2:2:10))
```

ararslan commented on base/abstractarray.jl:
```diff
@@ -388,6 +417,44 @@ julia> last([1; 2; 3; 4])
+julia> last(isodd, 2:2:10) |> isnothing
+true
+
+julia> last(iseven, [5; 3; 4; 2; 6; 9]) |> something
```
```suggestion
julia> something(last(iseven, [5; 3; 4; 2; 6; 9]))
```

non-Jedi committed:
Fixup code style in docstrings

From code review.

- Add ref links.
- Avoid using |>.
- Use , instead of ; between array elements.

Co-authored-by: Alex Arslan <ararslan@comcast.net>

non-Jedi commented on base/abstractarray.jl:
```diff
@@ -343,6 +343,35 @@ function first(itr)
+Get the first element of `coll` satisfying `predicate` wrapped in [`Some`](@ref).
+
+If no element of `coll` satisfies `predicate`, return `nothing`.
+
```
```suggestion

!!! compat "Julia 1.6"
    This method was added in Julia 1.6.

```

non-Jedi commented on base/abstractarray.jl:
```diff
@@ -388,6 +417,44 @@ julia> last([1; 2; 3; 4])
+Get the last element of `coll` satisfying `predicate` wrapped in [`Some`](@ref).
+
+If no element of `coll` satisfies `predicate`, return `nothing`.
+
```
```suggestion

!!! compat "Julia 1.6"
    This method was added in Julia 1.6.
    
```

non-Jedi reviewed (commented):
Should the docstrings instead be combined with the existing docstring for `first(coll)` as e.g.:

```julia
"""
    first([predicate=_ -> true,] coll)

...
"""
```

Also, I really appreciate the detailed code review. It would be great to get some feedback on the overall concept as well.

non-Jedi committed:
Add compat annotations to docstrings

mbauman commented on base/abstractarray.jl:
```diff
@@ -343,6 +343,38 @@ function first(itr)
+
+Get the first element of `coll` satisfying `predicate` wrapped in [`Some`](@ref).
+
+If no element of `coll` satisfies `predicate`, return `nothing`.
```
This is interesting — and deviates from the existing `first` which would throw an error.  This gives me a bit of pause, but at a minimum it means that we cannot combine these docstrings:

```Julia
julia> first(_->true, [])

julia> first([])
```

non-Jedi replied:
Good point on the docstrings.

mbauman commented on base/abstractarray.jl:
```diff
@@ -343,6 +343,38 @@ function first(itr)
+"""
+function first(predicate, itr)
+    for x in itr
+        predicate(x) && return Some(x)
```
I'm not thrilled about `Some` here — I don't think we have a single API in base that returns `Some`.  That said, this is unlike the other find/search APIs in that they return indices or `RegexMatch`es or values from some other limited type set that does not include `Nothing`.

non-Jedi replied:
I think I've convinced myself that at the very least this should be a separate function from `first`/`last`. Does `Iterators.first` (as suggested in OP) sound like a good name to you? In that case, we wouldn't have to create a predicate accepting version as it would compose with `Iterators.filter`.

In reality, I think all APIs that generically return an element of a collection but throw if the element isn't present should instead return `Union{Some{T},Nothing}`. This includes `first`, `last`, `getindex`, `reduce`, `foldl`, etc. This is obviously breaking; I should probably create a "Taking something seriously" issue to get feedback on that idea for julia 2.0.

mbauman commented on base/abstractarray.jl:
```diff
@@ -388,6 +420,47 @@ julia> last([1; 2; 3; 4])
+"""
+function last(predicate, itr)
+    out = nothing
+    for x in itr
```
I'd be inclined to not include this O(n) implementation  — I'd rather only support iterators that support `Iterators.Reverse` iteration.

non-Jedi replied:
Agreed. We don't include an O(n) implementation for regular `last` either.

mbauman reviewed (commented):
I'm not fully convinced about the functionality, but could be persuaded.

non-Jedi committed:
Create Iterators.first and Iterators.last

Instead of creating a version of Base.first and Base.last accepting a
predicate, create functions in the Iterators module. This allows easy
composition with Iterators.filter without having to accept a predicate
argument.

non-Jedi commented:
Could someone educate me on the history of `Some`? @mbauman correctly noted that no base functions currently return values wrapped in `Some`. It seems odd to have an exported type in base that's basically unused (except for a few internal reduction functions). Was it added to base to be a point of coordination for external packages? Is remnant of the days of `Nullable`? Is it purely so that `something(Some(_->nothing))` can return `nothing`?

non-Jedi commented:
I'd like to bump this.

The core functionality I'm trying to get into `Base` here is some function that allows getting the first element of a lazily filtered (possibly empty) collection without wrapping in a `try` block; a function returning `Union{Some{T},Nothing}` seems like the obvious way of doing this. I'm relatively ambivalent on how that should be spelled.

If the decision is that a function for getting the first element of a lazily filtered collection isn't necessary in `Base`, I'm happy to close this. If a different approach would be better, I'd be happy to work on this further.

rfourquet commented:
I like trying to solve the stated problem

>   there's not a good way of finding the first item matching a predicate in a non-indexable iterable [...]

* So my first comment: while having a non-throwing `first` seems like a good general solution, what about designing a function which solves directly the problem? e.g. `Iterators.findfirst(f, iter)` (which could for example return the "index" (a la `enumerate`) together with the found object).
 
* Regarding using `Some`, it looks like an appropriate use here even if it's currently seldom used elsewhere. Note this other [PR](#34821) implementing "2-argument `get` returning `Union{Nothing,Some}`".

* I don't like putting this functionality in `Iterators.first`. As far as I am aware, other public functions in `Iterators` create an iterator. Naming this one `Iterators.first` feels just like adding a random namespace in front of `first` just because `first` is already taken. What about `tryfirst` (by analogy to `tryparse` which doesn't throw (unlike `parse`) ?

non-Jedi commented:
> * So my first comment: while having a non-throwing `first` seems like a good general solution, what about designing a function which solves directly the problem? e.g. `Iterators.findfirst(f, iter)` (which could for example return the "index" (a la `enumerate`) together with the found object).

Not all iterables have meaningful indices (e.g. `readeach`), or is the idea that it would work exactly like `enumerate` and count the number of elements?

> * I don't like putting this functionality in `Iterators.first`. As far as I am aware, other public functions in `Iterators` create an iterator. Naming this one `Iterators.first` feels just like adding a random namespace in front of `first` just because `first` is already taken. What about `tryfirst` (by analogy to `tryparse` which doesn't throw (unlike `parse`) ?

I agree on the unsuitability of the `Iterators` name; mostly I was trying to put it next to `Iterators.filter` where I feel it's most useful. `tryfirst` is a pretty good name given the precedent of `tryparse`. Another possibility would be to introduce a new namespace/module where this could live as `first` alongside `Union{Some{T},Nothing}`-returning versions of `getindex`, `reduce`, etc.

non-Jedi commented:
So revisiting this in 2023, I think what I really want is for [Maybe.jl](https://github.com/JuliaPreludes/Maybe.jl) ([which didn't have `Maybe.first` at the time I authored this PR](https://github.com/JuliaPreludes/Maybe.jl/pull/32)) to be a standard library.

If there's not appetite for a change like this, I'm happy to update this PR to rename the function `tryfirst` as suggested by @rfourquet above. I think the specific case of `first` without throwing on empty is more important than the generalization Maybe.jl gives.

DilumAluthge commented:
This PR has had no activity for two years+, so I'll close it as stale.

@non-Jedi If you're still interested in working on this PR, let me know, and I can re-open it.
